---
title: "Isolate Assembly Report"
author: "`r Sys.getenv('USER')`"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    flexdashboard::flex_dashboard:
        orientation: rows
params:
  isolate_dir: "results/r62b14.hdr"
---

```{r setup}
knitr::opts_chunk$set(echo = FALSE)
library(tidyverse)
library(flexdashboard)
library(Biostrings)
```


```{r}
read_meta_json <- function(meta_json) {
	if (!file.exists(meta_json)) rlang::abort(str_glue("Cannot find {meta_json}"))
	json <- jsonlite::fromJSON(meta_json)
	json |> 
		enframe(name = NULL) |> 
		unnest_wider(value) |> 
		unnest_wider(c(assembly,org),names_sep = "_")
}


read_mlst_json <- function(mlst_json) {
	json <- jsonlite::fromJSON(mlst_json)
	json <- modify_in(json,c("mlst","user_input"),as_tibble)
	json <- modify_in(json,c("mlst","run_info"),as_tibble)
	json <- modify_in(
		json,
		c("mlst","results","allele_profile"),
		~ .x |>
			enframe(name = "mlst_gene") |>
			unnest_wider(value)
	)
	json$mlst
}


read_resfinder_json <- function(res_json) {
	json <- jsonlite::fromJSON(res_json)
	json$phenotypes <- json$phenotypes |> enframe() |> unnest_wider(value)
	json$seq_regions <- json$seq_regions |>
		enframe(name="seq_region_id") |>
		unnest_wider(value) |>
		mutate(contig_id = str_replace(query_id," .*",""))
	json$databases <- json$databases |> enframe(name="db_id") |> unnest_wider(value)
	json
}

read_plasmidfinder_json <- function(plf_json) {
	json <- jsonlite::fromJSON(plf_json,simplifyVector=FALSE)
	json <- modify_in(json,c("plasmidfinder","user_input"),as_tibble)
	json <- modify_in(json,c("plasmidfinder","run_info"),as_tibble)	
	json$plasmidfinder$results <- json |>
		pluck("plasmidfinder","results") |>
		modify_tree(leaf = ~ if (.=="No hit found") list() else .) |>
		enframe("db_lev1") |> 
		unnest_longer(value,indices_to = "db_lev2") |>
		unnest_longer(value,indices_to = "contig_name",values_to = "feature") |>
		unnest_wider(feature,names_sep = "_") |>
		mutate(contig_id = str_replace(feature_contig_name," .*",""))	|>
		relocate(contig_id)
	json$plasmidfinder
}

contigs <- function(fa) {
	tibble(contig_name = names(fa),contig_len=lengths(fa)) |>
		mutate(contig_id = str_replace(contig_name," .*",""))
}


read_isolate <- function(isolate_dir) {
	fa <- Biostrings::readDNAStringSet(fs::path(isolate_dir,"assembly.fasta"))
	list(
		fa = fa,
		meta = read_meta_json(fs::path(isolate_dir,"meta.json")),
		mlst = read_mlst_json(fs::path(isolate_dir,"mlst/data.json")),
		resfinder = read_resfinder_json(fs::path(isolate_dir,"resfinder/data.json")),
		plasmidfinder = read_plasmidfinder_json(fs::path(isolate_dir,"plasmidfinder/data.json")),
		contigs = contigs(fa),
		raw = list(
			resfinder = readLines(fs::path(isolate_dir,"resfinder/ResFinder_results.txt")) |> str_c(collapse="\n"),
			mlst = readLines(fs::path(isolate_dir,"mlst/results.txt")) |> str_c(collapse="\n"),
			plasmidfinder = readLines(fs::path(isolate_dir,"plasmidfinder/results.txt")) |> str_c(collapse="\n")
		)
	)
}
```


```{r}
x <- read_isolate(params$isolate_dir)
```



# Sidebar {.sidebar}

### Assembly
```{r}
valueBox(caption = "Assembly:",value=x$meta$assembly_id)
```




# Summary

Row
-------------------------

### Organism
```{r}
valueBox(x$meta$org_org_name,"Organism name",icon = "glyphicon glyphicon-user")
```

Row
-------------------------

### Total length
```{r}
valueBox(sprintf("%s bp",format(sum(x$contigs$contig_len),big.mark="'")),"Total length",icon='glyphicon glyphicon-resize-horizontal')
```

### Number of contig
```{r}
valueBox(nrow(x$contigs),"Contigs",icon='glyphicon glyphicon-scissors')
```

### ANI
```{r}
gauge(as.numeric(x$meta$org_org_ani), min = 80, max = 100, symbol = '%', gaugeSectors(
  success = c(95, 100), warning = c(90, 95), danger = c(80, 90)
))
```

Row
-------------------------

### MLST
```{r}
valueBox(x$mlst$results$sequence_type,"MLST",icon = "glyphicon glyphicon-tags")
```




# Assembly QC

Row
-------------------------
### Total length
```{r}
valueBox(sprintf("%s bp",format(sum(x$contigs$contig_len),big.mark="'")),"Total length",icon='glyphicon glyphicon-resize-horizontal')
```

### Number of contig
```{r}
valueBox(nrow(x$contigs),"Contigs",icon='glyphicon glyphicon-scissors')
```

### N50
```{r}
valueBox(sprintf("%s bp",format(Biostrings::N50(x$contigs$contig_len),big.mark="'")),"N50",icon='glyphicon glyphicon-equalizer')
```


Row
-------------------------

### Contig size distribution

```{r}
x$contigs |>
	mutate(contig_id = fct_reorder(contig_id,contig_len,mean)) |>
	ggplot() + 
		geom_segment(aes(x=0,y=contig_id,xend=contig_len,yend=contig_id)) + 
		scale_x_continuous(labels = scales::label_number(scale_cut = scales::cut_short_scale())) +
		xlab("Contig size (bp)") +
		theme_bw()
```

### Contig topology

Row
-------------------------

### Contig coverage

### GC content





# Contigs


## Summary

### Resistances
```{r}
x$resfinder$seq_regions |>
	left_join(x$contigs,by="contig_id") |>
	mutate(contig_id = str_glue("{contig_id} - {contig_len}bp")) |>
	mutate(contig_id = fct_reorder(contig_id,contig_len,mean)) |>
	arrange(desc(contig_id)) |>
	select(contig_id,resistance=name,identity,coverage) |>
	arrange(desc(contig_id)) |>
	group_by(contig_id) |>
	gt::gt(row_group_as_column=TRUE,caption = "Detected resistances per contig") |>
	gt::tab_stubhead("contig - len - topo")  |>
	gt::tab_style(list(gt::cell_text(weight="bold"),gt::cell_fill()),locations=gt::cells_column_labels()) |>
	gt::tab_style(list(gt::cell_text(weight="bold"),gt::cell_fill()),locations=gt::cells_stubhead()) |>
	gt::tab_style(list(gt::cell_fill(color = "green")),locations=gt::cells_body(identity,rows = identity>=100)) |>
	gt::tab_style(list(gt::cell_fill(color = "green")),locations=gt::cells_body(coverage,rows = coverage>=100)) |>
	gt::fmt_number(c(identity,coverage),pattern = "{x}%")
```



### Plasmids

```{r}
x$plasmidfinder$results |>
	left_join(x$contigs,by=c("contig_id")) |>
	mutate(contig_id = str_glue("{contig_id} - {contig_len}bp")) |>
	mutate(contig_id = fct_reorder(contig_id,contig_len,mean)) |>
	arrange(desc(contig_id),desc(feature_identity)) |>
	select(contig_id,plasmid_type=feature_plasmid,identity=feature_identity,coverage=feature_coverage) |>
	group_by(contig_id) |>
	gt::gt(row_group_as_column=TRUE,caption = "Detected plasmid feature per contig") |>
	gt::tab_stubhead("contig - length") |>
	gt::tab_style(list(gt::cell_text(weight="bold"),gt::cell_fill()),locations=gt::cells_column_labels()) |>
	gt::tab_style(list(gt::cell_text(weight="bold"),gt::cell_fill()),locations=gt::cells_stubhead()) |>
	gt::tab_style(list(gt::cell_fill(color = "green")),locations=gt::cells_body(identity,rows = identity>=100)) |>
	gt::tab_style(list(gt::cell_fill(color = "green")),locations=gt::cells_body(coverage,rows = coverage>=100)) |>
	gt::fmt_number(c(identity,coverage),pattern = "{x}%")
```



# Raw outputs

## {.tabset}

### MLST
```{r,results='asis'}
htmltools::tags$pre(
	style = "max-height: 700px; overflow-y: auto; overflow-x: auto; white-space: pre; display: block; width: 100%;",
	htmltools::tags$code(
		style = "display: block; width: max-content;",
		x$raw$mlst
	)
)
```

### Resfinder
```{r,results='asis'}
htmltools::tags$pre(
	style = "max-height: 700px; overflow-y: auto; overflow-x: auto; white-space: pre; display: block; width: 100%;",
	htmltools::tags$code(
		style = "display: block; width: max-content;",
		x$raw$resfinder
	)
)
```

### Plasmidfinder
```{r,results='asis'}
htmltools::tags$pre(
	style = "max-height: 700px; overflow-y: auto; overflow-x: auto; white-space: pre; display: block; width: 100%;",
	htmltools::tags$code(
		style = "display: block; width: max-content;",
		x$raw$plasmidfinder
	)
)
```


