---
title: "Multiple Plasmid Alignment Visualisation"
output: html_document
params:
  out_dir: "out/fig"
---

```{r,include=FALSE}
library(Rsamtools)
library(tidyverse)
source("src/lib_res_pipeline.R")
source("src/lib_ggbrowser.R")
```


```{bash eval=FALSE}
docker run --rm -v ./:/cwd \
  --workdir /cwd ncbi/blast:2.13.0 \
  blastn -max_hsps 1000 -perc_identity 50 -evalue 1e-3 \
    -query out/blaCTX-M-15_IncHI2A_plasmids.fasta \
    -subject out/blaCTX-M-15_IncHI2A_plasmids.fasta \
    -outfmt "6 std qlen slen" \
    -out out/blaCTX-M-15_IncHI2A_plasmids.self.blastn
```

```{r,include=FALSE}
fa <- FaFile("data/figS1_review/ext_fig1_plasmids_repA_review.fasta")
indexFa(fa)
aln <- read_blast6("data/figS1_review/ext_fig1_plasmids_repA_review.self.blastn",seqinfo = seqinfo(fa))
gff <- list.files("data/figS1_review",".gff$",full.names = TRUE,recursive=TRUE)
gff <- lapply(gff,function(f) {
  print(f)
  gff <- rtracklayer::import.gff(f)
  if (all(seqlevels(gff) %in% seqlevels(fa))) {
    gff <- rtracklayer::import.gff(f,genome=seqinfo(fa))
  } else {
    gff <- gff[integer(0)]    
  }
  gff$Label <- NULL;gff$Label <- rep_len(NA,length(gff))
  
  if (!is.null(gff$blue)) gff$Label[is.na(gff$Label)] <- gff$blue[is.na(gff$Label)]
  if (!is.null(gff$red)) gff$Label[is.na(gff$Label)] <- gff$red[is.na(gff$Label)]
  if (!is.null(gff$Plasmid)) gff$Label[is.na(gff$Label)] <- gff$Plasmid[is.na(gff$Label)]
  if (!is.null(gff$"Resistance gene")) gff$Label[is.na(gff$Label)] <- gff$"Resistance gene"[is.na(gff$Label)]
  if (!is.null(gff$Name)) gff$Label[is.na(gff$Label)] <- gff$Name[is.na(gff$Label)]
  if (is.null(gff$blue)) gff$blue <- rep_len(NA,length(gff))
  if (is.null(gff$red)) gff$red <- rep_len(NA,length(gff))
  mcols(gff) <- mcols(gff)[c("source","type","Label","red","blue")]
  gff
}) |> 
  setNames(gff) |>
  GRangesList() |>
  stack()
seqinfo(gff,match(seqlevels(fa),seqlevels(gff))) <- seqinfo(fa)


gff <- gff[(gff$source %in% c("resfinder","plasmidfinder")) | (gff$type %in% c("CDS","unsure","insertion_sequence"))]
gff$Label <- sub(" CDS$","",gff$Label)
gff$Label[gff$Label %in% c("hypothetical protein")] <- NA
#gff$Label[!(gff$type %in% c("unsure","insertion_sequence")) & is.na(gff$blue) & is.na(gff$red) & !(gff$source %in% c("resfinder","plasmidfinder"))] <- NA
gff$Label <- sub("_[0-9]*$","",gff$Label)
gff$source <- fct_relevel(gff$source,"resfinder","plasmidfinder","mge_finder")

gff <- gff[order(gff$source,gff$Label)]
gff <- gff[!duplicated(gff)]

gff[gff$source=="resfinder" & gff$Label=="blaOXA-48"]

#gff$fill <- "#FFFFFF"
gff$fill <- "#F3EAE4"
gff$fill[!is.na(gff$red)] <- "#E41A1C"
gff$fill[gff$source %in% "resfinder"] <- "#E41A1C"
gff$fill[!is.na(gff$blue)] <- "#3333FF"
gff$fill[gff$source %in% "mge_finder"] <- "#3333FF"
gff$fill[gff$source %in% "plasmidfinder"] <- "#FFFF33"
gff <- gff[order(gff$fill,decreasing = TRUE)]
```



# Make a pairwise comparison plot
```{r}
#ggdotplot(aln)
```


# Make consecutive sequence comparison plot
```{r fig.height=15, fig.width=13}
keep_maximal <- function(aln) {
    h1 <- findOverlaps(unstrand(aln$subject),type = "within")
    h2 <- findOverlaps(unstrand(aln$query),type = "within")
    h <- BiocGenerics::intersect(h1,h2)
    h <- h[subjectHits(h) != queryHits(h)]
    aln[-queryHits(h),]
}

local({
  ALN <- aln |>
    subset(pct_ident >= 90) |>
    subset(align_len >= 200) 
  ALN <- ALN[abs(as.integer(seqnames(ALN$query))-as.integer(seqnames(ALN$subject)))==1,]
  # head(ALN[order(-ALN$align_len),],10)
  ALN <- keep_maximal(ALN)

  gff$bin <- local({
    strand(gff)[strand(gff)=="*"] <- "+"
    disjointBins(gff) - 0.5
  })
  #gff$bin <- ifelse(strand(gff) %in% "-",-gff$bin,+gff$bin)
  gff$bin <- 0
  gff_poly <- gff_polygon_data(gff,arrow_head_len=200)
  gff_poly <- cbind(gff_poly,mcols(gff)[gff_poly$i,c("source","Label","bin","fill")])
  gff_poly$y <- gff_poly$y*0.1 + gff_poly$bin * 0.11 + as.integer(seqnames(gff))[gff_poly$i]

  aln_poly <- self_blastn_polygon_data(ALN)
  aln_poly$fill <- colorRampPalette(RColorBrewer::brewer.pal(9,"Blues"))(30)[as.integer(cut(aln_poly$pct_ident,seq(80,100,length.out=31)))]
  aln_poly$fill <- "#7BCCC4"
  
  ggplot() + 
    geom_polygon(aes(x=x,y=y+dy*0.1,group=idx,fill=fill),alpha=0.5,color="grey",data=aln_poly) +
    geom_segment(aes(x=1,xend=seqlengths,y=seq_along(seqlengths),yend=after_stat(y)),data=as.data.frame(seqinfo(gff))) +
    geom_polygon(aes(x=x,y=y,group=i,fill=fill),data=gff_poly,color="darkgrey") +
    #ggrepel::geom_text_repel(aes(x=(start+end)/2,y=as.integer(seqnames)+bin*0.11,label=Label),data=as.data.frame(gff[gff$source%in%c("resfinder","plasmidfinder","mge_finder")]),size=2,hjust=0.5,max.overlaps = +Inf,min.segment.length = 0.2) +
    ggrepel::geom_text_repel(
      aes(x=(start+end)/2,y=as.integer(seqnames)+bin*0.11,label=Label),
      data = dplyr::filter(as.data.frame(gff),(seqnames %in% "CP047704.1") | (source!="Prodigal:002006")),
      size=2,hjust=0.5,max.overlaps = +Inf,min.segment.length = 0.2,box.padding = 0.5) +
    scale_fill_identity() +
    scale_x_continuous(breaks = c(1,seq(10000,max(seqlengths(gff)),by=10000)),labels = scales::label_number(scale_cut = scales::cut_short_scale())) +
    #scale_y_continuous(breaks=seq_along(seqlevels(gff)),labels = seqlevels(gff)) +
    scale_y_reverse(breaks=seq_along(seqlevels(gff)),labels = seqlevels(gff)) +
    xlab("") + ylab("") +
    theme(
      panel.grid = element_blank(),panel.background = element_blank(),
      legend.position="none",axis.ticks.y = element_blank())
})
ggsave(file.path(params$out_dir,"consecutive_sequence_comparison_ext_fig1_clean.pdf"),limitsize = FALSE)
```





